### PR TITLE
Fix #83 - introduce global timeouts for retry policies

### LINKED ISSUE #83: Unbound retry policy continues retrying despite the finished original future
Steps to reproduce:

1. Configure a `RetryingTarantoolTupleClient` with `TarantoolRequestRetryPolicies.unbound()`
2. Execute an operation (e.g. `callForSingleResult`) with explicit timeout (using `get(time, timeUnit))`

Actual behavior:

Retries continue to be performed despite that future finished with timeout

Expected behavior:

No more retries are performed after the request future is finished with timeout

### PR DESCRIPTION
**Pull request by vrogach2020**
Fixes #83 
Introduce operationTimeout parameter to configure exact timeout until retry attempts will be made.

### CODE REVIEW

File: src/main/java/io/tarantool/driver/retry/RequestRetryPolicy.java [hunk 1]
```diff
@@ -33,45 +34,41 @@
...
+     * @return timeout, in milliseconds
      */
-    default long getOperationTimeout() {
+    default long getRequestTimeout() {
         return TimeUnit.HOURS.toMillis(1);
     }
 
     /**
      * Wrap a generic operation taking an arbitrary number of arguments and returning a {@link CompletableFuture}.
-     * Each operation attempt is limited with a timeout returned by {@link #getOperationTimeout()}
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/retry/RequestRetryPolicy.java:46]: I think this line should be left here

File: src/main/java/io/tarantool/driver/retry/TarantoolRequestRetryPolicies.java [hunk 2]
```diff
@@ -117,9 +164,10 @@ public RequestRetryPolicy create() {
          */
         public static class Builder<T extends Function<Throwable, Boolean>> {
 
-            private long timeout = TimeUnit.HOURS.toMillis(1); //ms
+            private long requestTimeout = TimeUnit.HOURS.toMillis(1); //ms
             private long delay; //ms
             private final T callback;
+            private long operationTimeout = TimeUnit.DAYS.toMillis(30);
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/retry/TarantoolRequestRetryPolicies.java:170]: Seems too big, I think 1 hour is enough for default.

File: src/main/java/io/tarantool/driver/retry/TarantoolRequestRetryPolicies.java [hunk 3]
```diff
@@ -173,38 +226,38 @@ public Builder(T callback) {
...
+         * @param exceptionCheck function checking whether the given exception may be retried
          */
-        public AttemptsBoundRetryPolicy(int attempts, long timeout, long delay, T callback) {
+        public AttemptsBoundRetryPolicy(int attempts, long requestTimeout, long delay, T exceptionCheck) {
             Assert.state(attempts >= 0, "Attempts must be greater or equal than 0!");
-            Assert.state(timeout >= 0, "Timeout must be greater or equal than 0!");
+            Assert.state(requestTimeout >= 0, "Timeout must be greater or equal than 0!");
             Assert.state(delay >= 0, "Timeout must be greater or equal than 0!");
-            Assert.notNull(callback, "Callback must not be null!");
+            Assert.notNull(exceptionCheck, "Callback must not be null!");
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/retry/TarantoolRequestRetryPolicies.java:250]: "Callback" -> "Exception checking callback"

File: src/main/java/io/tarantool/driver/retry/package-info.java [hunk 4]
```diff
@@ -0,0 +1,4 @@
+/**
+ * Network errors handling functionality.
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/retry/package-info.java:2]: Not only networks.
Request retrying functionality?

File: src/test/java/io/tarantool/driver/integration/ClusterConnectionIT.java [hunk 5]
```diff
@@ -178,6 +178,13 @@ void testMultipleConnectionsReconnect_retryableRequestShouldNotFail() throws Exc
         CompletableFuture<Boolean> request2 = client.callForSingleResult(
                 "long_running_function", Collections.singletonList(0.5), Boolean.class);
 
+        //wait until connections are created
+        int timeoutCounter = 10;
```
Comment (akudiyar) [inline src/test/java/io/tarantool/driver/integration/ClusterConnectionIT.java:182]: Seems that we need to inspect the TC/Docker API and provide some health checks in the cluster for reliable waiting for start

File: src/test/java/io/tarantool/driver/integration/RetryingTarantoolTupleClientIT.java [hunk 6]
```diff
@@ -53,8 +64,84 @@ void testSuccessAfterSeveralRetries() throws Exception {
...
+        CompletableFuture<List<?>> f = retryingClient
+                .call("long_running_function", Collections.singletonList(20));
+
+        assertThrows(TimeoutException.class, () -> f.get(500, TimeUnit.MILLISECONDS));
+        assertFalse(f.isDone());
+
+        long counterAfterCompletion = counter.get();
+        assertTrue(counterAfterCompletion > 1,
+                "Request was not executed by policy.");
+        Thread.sleep(1500);
```
Comment (akudiyar) [inline src/test/java/io/tarantool/driver/integration/RetryingTarantoolTupleClientIT.java:138]: I think we can operate with smaller numbers for not blocking the tests for 2 seconds. 10-20-30 ms might be enough

File: src/test/java/io/tarantool/driver/integration/SharedCartridgeContainerStub.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is a debugging stub that mimics a tarantool
+ * testcontainer but works with a running cluster.
+ *
+ * @author Vladimir Rogach
+ */
+abstract class SharedCartridgeContainerStub {
```
Comment (akudiyar) [inline src/test/java/io/tarantool/driver/integration/SharedCartridgeContainerStub.java:13]: Make this class not abstract, but a service that is enabled via Maven profiles (or system variables).

**Review by akudiyar — CHANGES_REQUESTED** [review]

File: src/test/java/io/tarantool/driver/integration/ClusterConnectionIT.java (hunk 5 above)
Comment (vrogach2020) [reply src/test/java/io/tarantool/driver/integration/ClusterConnectionIT.java:182]: dropped, need to investigate healthchecks

File: src/test/java/io/tarantool/driver/integration/SharedCartridgeContainerStub.java (hunk 7 above)
Comment (vrogach2020) [reply src/test/java/io/tarantool/driver/integration/SharedCartridgeContainerStub.java:13]: Dropped by now, need to do some big refactoring for this

File: src/test/java/io/tarantool/driver/integration/SharedCartridgeContainer.java [hunk 8]
```diff
@@ -10,13 +9,12 @@
 
     private static final Logger logger = LoggerFactory.getLogger(SharedCartridgeContainer.class);
 
-    @ClassRule
```
Comment (akudiyar) [inline src/test/java/io/tarantool/driver/integration/SharedCartridgeContainer.java:13]: Does this change not broke anything?

Comment (vrogach2020) [reply src/test/java/io/tarantool/driver/integration/SharedCartridgeContainer.java:13]: No, it seems junit5 is used everywhere

**Review by akudiyar — APPROVED** [review]

**Pull request merged**